### DISCUSSION TITLE
How to change the runtime module's content

### DISCUSSION THREAD

**Discussion by tmkx** [web-infra-dev/rspack#5269] [Q&A]
## Background
I want to build a chrome extension using rspack but encountered some problems:

1. In [content-scripts](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts), the style tags may exist under ShadowRoot, [`document.getElementsByTagName("link")`](https://github.com/web-infra-dev/rspack/blob/1fcafc3fa6499ead35577d9bd6f8bc77ad6254bc/crates/rspack_plugin_runtime/src/runtime_module/runtime/css_loading.js#L9) cannot find the style elements, leading to a non-working HMR.
2. A specified entry runs in the Service Worker environment ([background](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers)), and it is unable to load JS code dynamically. It is hoped that the `webpack/runtime/load` runtime script code can be customized, so that I can perform a reload action when the entry receives a HMR message.
3. The public path of content-script files is runtime-evaluated.

## Webpack demo
It's a demo that shows how to customize the runtime public path:
```ts
export class ContentScriptPublicPathPlugin extends ContentScriptBasePlugin {
  apply(compiler: webpack.Compiler) {
    compiler.hooks.thisCompilation.tap('ContentScriptPublicPathPlugin', (compilation) => {
      compilation.hooks.runtimeModule.tap('ContentScriptPublicPathPlugin', (module, chunk) => {
        if (chunk.getEntryOptions()?.name !== 'content-script') return;
        if (module.name === 'publicPath') patchPublicPathRuntimeModule(module, compiler);
      });
    });
  }
}

function patchPublicPathRuntimeModule(
  module: webpack.RuntimeModule,
  { webpack, webpack: { RuntimeGlobals } }: webpack.Compiler
) {
  module.name = 'content-script publicPath';
  module.generate = function (this: webpack.RuntimeModule) {
    return webpack.Template.asString([`${RuntimeGlobals.publicPath} = chrome.runtime.getURL('/');`]);
  };
}
```

**Comment by hardfist** [comment]
this is a reasonable feature request, we will consider support it

**Comment by hardfist** [comment]
@tmkx runtimeModule is supported in 0.5.3, can you give it a try

**Reply by tmkx** [reply]
absolutely, salute for the iteration speed 🫡

**Reply by tmkx** [reply]
https://github.com/web-infra-dev/rspack/issues/5571

I have only encountered one issue so far
